File: hydra-clojure/src/main/clojure/hydra/lib/equality.clj
(ns hydra.lib.equality)

(defn generic-compare [a b]
  (cond
    (= a b) 0
    (and (number? a) (number? b)) (compare a b)
    (and (string? a) (string? b)) (compare a b)
    (and (char? a) (char? b)) (compare (int a) (int b))
    (and (keyword? a) (keyword? b)) (compare a b)
    (and (boolean? a) (boolean? b)) (compare a b)
    (and (sequential? a) (sequential? b))
    (let [len-a (count a) len-b (count b)]
      (if (not= len-a len-b)
        (compare len-a len-b)
        (loop [ra (seq a) rb (seq b)]
          (if (nil? ra) 0
              (let [c (generic-compare (clojure.core/first ra) (clojure.core/first rb))]
                (if (not= c 0) c (recur (next ra) (next rb))))))))
    (and (nil? a) (nil? b)) 0
    (nil? a) -1
    (nil? b) 1
    :else (compare (pr-str a) (pr-str b))))

;; compare :: a -> a -> Comparison
;; Returns a Comparison union: (list :less_than nil), (list :equal_to nil), (list :greater_than nil)
(def hydra_lib_equality_compare
  (fn [a] (fn [b]
    (let [c (generic-compare a b)]
      (cond
        (neg? c) (list :less_than nil)
        (zero? c) (list :equal_to nil)
        :else (list :greater_than nil))))))

;; equal :: a -> a -> Bool
(def hydra_lib_equality_equal
  (fn [a] (fn [b] (= a b))))

;; gt :: a -> a -> Bool
(def hydra_lib_equality_gt
  (fn [a] (fn [b] (> (generic-compare a b) 0))))

;; gte :: a -> a -> Bool
(def hydra_lib_equality_gte
  (fn [a] (fn [b] (>= (generic-compare a b) 0))))

;; identity :: a -> a
(def hydra_lib_equality_identity
  (fn [x] x))

;; lt :: a -> a -> Bool
(def hydra_lib_equality_lt
  (fn [a] (fn [b] (< (generic-compare a b) 0))))

;; lte :: a -> a -> Bool
(def hydra_lib_equality_lte
  (fn [a] (fn [b] (<= (generic-compare a b) 0))))

;; max :: a -> a -> a
(def hydra_lib_equality_max
  (fn [a] (fn [b] (if (>= (generic-compare a b) 0) a b))))

;; min :: a -> a -> a
(def hydra_lib_equality_min
  (fn [a] (fn [b] (if (<= (generic-compare a b) 0) a b))))

File: heads/typescript/src/main/typescript/hydra/lib/equality.ts
// Hand-written runtime: hydra.lib.equality primitives.
//
// Hydra equality is structural — JS reference equality only works for primitives.
// For objects we use JSON.stringify as a pragmatic fallback; this matches the
// behavior of Hydra's encode-and-compare semantics for serializable data.

export const equal = <A>(a: A, b: A): boolean => {
  if (Object.is(a, b)) return true;
  if (typeof a !== "object" || typeof b !== "object" || a === null || b === null) return false;
  return JSON.stringify(a) === JSON.stringify(b);
};

// Unwrap a Hydra Term value to its underlying scalar for ordered
// comparison. The kernel test primitives often compare Term-encoded
// integers/floats/strings; raw structural JSON comparison fails for
// these (e.g. "...1}}}" lex-compares wrong against "...100}}}").
const unwrapForOrdering = (x: unknown): unknown => {
  if (x === null || typeof x !== "object") return x;
  const o = x as { tag?: string; value?: unknown };
  // Term_literal → unwrap to the literal value.
  if (o.tag === "literal" && o.value && typeof o.value === "object") {
    const lit = o.value as { tag?: string; value?: unknown };
    if (lit.tag === "string" || lit.tag === "binary") return lit.value;
    if (lit.tag === "boolean") return lit.value;
    // Integer/float wrap one more level.
    if (lit.tag === "integer" || lit.tag === "float") {
      const inner = lit.value as { tag?: string; value?: unknown };
      return inner?.value;
    }
    if (lit.tag === "decimal") return lit.value;
  }
  // Wrap (newtype Name etc.): { value: <inner> }.
  if (Object.keys(o).length === 1 && "value" in o) return unwrapForOrdering(o.value);
  return x;
};

export const lt = <A>(a: A, b: A): boolean => {
  const ua = unwrapForOrdering(a);
  const ub = unwrapForOrdering(b);
  if (typeof ua === "number" && typeof ub === "number") return ua < ub;
  if (typeof ua === "string" && typeof ub === "string") return ua < ub;
  if (typeof ua === "bigint" && typeof ub === "bigint") return ua < ub;
  // Fallback: original structural comparison.
  return JSON.stringify(a) < JSON.stringify(b);
};

export const lte = <A>(a: A, b: A): boolean => equal(a, b) || lt(a, b);
export const gt = <A>(a: A, b: A): boolean => lt(b, a);
export const gte = <A>(a: A, b: A): boolean => equal(a, b) || gt(a, b);

// Hydra `compare` returns a hydra.util.Comparison value:
//   { tag: "lessThan" } | { tag: "equalTo" } | { tag: "greaterThan" }
// The kernel pattern-matches on the tag (see hydra/show/util.ts), so
// the runtime value must use this discriminated-union shape — not a
// raw -1/0/1 number.
export type Comparison =
  | { readonly tag: "lessThan" }
  | { readonly tag: "equalTo" }
  | { readonly tag: "greaterThan" };

export const compare = <A>(a: A, b: A): Comparison => {
  if (equal(a, b)) return { tag: "equalTo" } as const;
  return lt(a, b) ? { tag: "lessThan" } as const : { tag: "greaterThan" } as const;
};

export const identity = <A>(a: A): A => a;

export const min = <A>(a: A, b: A): A => lt(a, b) ? a : b;
export const max = <A>(a: A, b: A): A => lt(b, a) ? a : b;
